certbot-dns-route53/certbot_dns_route53/dns_route53.py

```diff
@@ -42,15 +42,30 @@ def more_info(self):  # pylint: disable=missing-docstring,no-self-use
     def _setup_credentials(self):
         pass
 
-    def _perform(self, domain, validation_domain_name, validation):
-        try:
-            change_id = self._change_txt_record("UPSERT", validation_domain_name, validation)
+    def perform(self, achalls):
+        self._attempt_cleanup = True
 
-            self._wait_for_change(change_id)
+        try:
+            change_ids = [
+                self._change_txt_record("UPSERT",
+                  achall.validation_domain_name(achall.domain),
+                  achall.validation(achall.account_key))
+                for achall in achalls
+            ]
+
+            for change_id in change_ids:
+                self._wait_for_change(change_id)
         except (NoCredentialsError, ClientError) as e:
             logger.debug('Encountered error during perform: %s', e, exc_info=True)
             raise errors.PluginError("\n".join([str(e), INSTRUCTIONS]))
 
+        # Sleep for at least the TTL, to ensure that any records cached by
+        # the ACME server after previous validation attempts are gone. In
+        # most cases we'll need to wait at least this long for the Route53
+        # records to propagate, so this doesn't delay us much.
+        time.sleep(self.ttl)
```

Comment:
Disclaimer: it has a been a while since I've looked at this, and I don't have a test system handy anymore.

I'm not sure this sleep is necessary. In most cases, I'd expect validation to occur very infrequently relative to the TTL (on the order of months vs. the order of minutes). Because records are deleted at the end of each execution, I wouldn't expect records from a previous challenge to still be cached (and if the absence of a record were being cached, that would be [determined by the TTL on the SOA record](https://tools.ietf.org/html/rfc2308), not our `self.ttl` value here, which seems like something that would cause major problems for our workflow since SOA TTLs might be very high).

Once `_wait_for_change` has returned for each `change_id`, it seems like we should be good to ask the server to validate the challenges.

There are some edge cases around, for example, having a pool of machines behind a load balancer which each order certificates for the same domain on a similar schedule. However, that seems like a less common use case and retrying in cases like that is something we should be handling in a generic way (since the issue could occur with any DNS provider).

Reply:
Updated to remove the unnecessary sleep, thanks.